### ISSUE TITLE
_TestTCPForwarding test case failure on Fedora 36

### ISSUE THREAD

**Issue by gsauthof** [ronf/asyncssh#446]
Running the asyncssh 2.9.0 test suite on Fedora 36/Rawhide yields 4 failures in TCP forwarding tests:

```
======================================================================
FAIL: test_forward_remote_port (tests.test_forward._TestAsyncTCPForwarding)
Test forwarding of a remote port
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/util.py", line 81, in async_wrapper
    return self.loop.run_until_complete(coro(self, *args, **kwargs))
  File "/usr/lib64/python3.10/asyncio/base_events.py", line 641, in run_until_complete
    return future.result()
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 687, in test_forward_remote_port
    await self._check_local_connection(listener.get_port())
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 287, in _check_local_connection
    await self._check_echo_line(reader, writer, delay=delay)
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 243, in _check_echo_line
    self.assertEqual(line, result)
AssertionError: b'140735368876512\n' != b''
======================================================================
FAIL: test_forward_remote_specific_port (tests.test_forward._TestAsyncTCPForwarding)
Test forwarding of a specific remote port
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/util.py", line 81, in async_wrapper
    return self.loop.run_until_complete(coro(self, *args, **kwargs))
  File "/usr/lib64/python3.10/asyncio/base_events.py", line 641, in run_until_complete
    return future.result()
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 708, in test_forward_remote_specific_port
    await self._check_local_connection(listener.get_port())
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 287, in _check_local_connection
    await self._check_echo_line(reader, writer, delay=delay)
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 243, in _check_echo_line
    self.assertEqual(line, result)
AssertionError: b'140735368876896\n' != b''
======================================================================
FAIL: test_forward_remote_port (tests.test_forward._TestTCPForwarding)
Test forwarding of a remote port
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/util.py", line 81, in async_wrapper
    return self.loop.run_until_complete(coro(self, *args, **kwargs))
  File "/usr/lib64/python3.10/asyncio/base_events.py", line 641, in run_until_complete
    return future.result()
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 687, in test_forward_remote_port
    await self._check_local_connection(listener.get_port())
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 287, in _check_local_connection
    await self._check_echo_line(reader, writer, delay=delay)
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 243, in _check_echo_line
    self.assertEqual(line, result)
AssertionError: b'140735368879008\n' != b''
======================================================================
FAIL: test_forward_remote_specific_port (tests.test_forward._TestTCPForwarding)
Test forwarding of a specific remote port
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/util.py", line 81, in async_wrapper
    return self.loop.run_until_complete(coro(self, *args, **kwargs))
  File "/usr/lib64/python3.10/asyncio/base_events.py", line 641, in run_until_complete
    return future.result()
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 708, in test_forward_remote_specific_port
    await self._check_local_connection(listener.get_port())
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 287, in _check_local_connection
    await self._check_echo_line(reader, writer, delay=delay)
  File "/builddir/build/BUILD/asyncssh-2.9.0/tests/test_forward.py", line 243, in _check_echo_line
    self.assertEqual(line, result)
AssertionError: b'140735368887888\n' != b''
----------------------------------------------------------------------
```

cf. https://kojipkgs.fedoraproject.org//work/tasks/9919/82419919/build.log

**Comment by gsauthof** [comment]
PS: They are also failing on my local machine, which is running Fedora 35.

**Comment by ronf** [comment]
Thanks for the report. Is this failure new in 2.9.0? I don't have a recent Fedora system here to test against, but I can try to spin one up. I've seen and fixed similar errors in this code previously, though, and it looks like maybe I missed a couple of spots. Could you try the following patch and see if it helps?

```python
diff --git a/tests/test_forward.py b/tests/test_forward.py
index bdceb6e..99e0f50 100644
--- a/tests/test_forward.py
+++ b/tests/test_forward.py
@@ -677,7 +677,7 @@ class _TestTCPForwarding(_CheckForwarding):
     async def test_forward_remote_port(self):
         """Test forwarding of a remote port"""

-        server = await asyncio.start_server(echo, None, 0,
+        server = await asyncio.start_server(echo, '', 0,
                                             family=socket.AF_INET)
         server_port = server.sockets[0].getsockname()[1]

@@ -693,7 +693,7 @@ class _TestTCPForwarding(_CheckForwarding):
     async def test_forward_remote_specific_port(self):
         """Test forwarding of a specific remote port"""

-        server = await asyncio.start_server(echo, None, 0,
+        server = await asyncio.start_server(echo, '', 0,
                                             family=socket.AF_INET)
         server_port = server.sockets[0].getsockname()[1]

```

There seems to be an inconsistency between different flavors of UNIX about the behavior in Python when passing in `None` vs. an empty string as a listening address on a socket.

**Comment by gsauthof** [comment]
Yes, this is new. With the previous version, 2.8.1, no test cases failed.
There was also a mass package rebuild with the start of Fedora 36 development, where all the tests succeeded.

The patch doesn't help:

```
$ pytest tests -k TCPForwarding
...
============================================================= short test summary info 
FAILED tests/test_forward.py::_TestTCPForwarding::test_forward_remote_port - AssertionError: b'139743820507280\n' != b''
FAILED tests/test_forward.py::_TestTCPForwarding::test_forward_remote_specific_port - AssertionError: b'139743821085952\n' != b''
FAILED tests/test_forward.py::_TestAsyncTCPForwarding::test_forward_remote_port - AssertionError: b'139743788262704\n' != b''
FAILED tests/test_forward.py::_TestAsyncTCPForwarding::test_forward_remote_specific_port - AssertionError: b'139743821072224\n' != b''
============================================ 4 failed, 78 passed, 1360 deselected in 292.19s (0:04:52) 
```

**Comment by ronf** [comment]
Ok - thanks for the quick turnaround on that.. It was worth a shot!

I'll try and get an environment set up to attempt to reproduce this.

**Comment by ronf** [comment]
Ok - I was able to reproduce this on an Ubuntu system here as it turns out. Could you see if the following fix works on your Fedora setup?

```patch
index bdceb6e..321bd61 100644
--- a/tests/test_forward.py
+++ b/tests/test_forward.py
@@ -682,8 +682,8 @@ class _TestTCPForwarding(_CheckForwarding):
         server_port = server.sockets[0].getsockname()[1]

         async with self.connect() as conn:
-            async with conn.forward_remote_port('', 0,
-                                                '', server_port) as listener:
+            async with conn.forward_remote_port(
+                    '', 0, '127.0.0.1', server_port) as listener:
                 await self._check_local_connection(listener.get_port())

         server.close()
@@ -703,8 +703,8 @@ class _TestTCPForwarding(_CheckForwarding):
         sock.close()

         async with self.connect() as conn:
-            async with conn.forward_remote_port('', remote_port,
-                                                '', server_port) as listener:
+            async with conn.forward_remote_port(
+                    '', remote_port, '127.0.0.1', server_port) as listener:
                 await self._check_local_connection(listener.get_port())

         server.close()
```

This appears to only be a flaw in the tests, and not in AsyncSSH's main code. Looking at the diff from 2.8.1, the function which checks the forwarding was changed from using an empty string as a destination address to using '127.0.0.1' because of issues I found on other platforms, but the call to forward_remote_port() was still using an empty string as the local address to forward to, and this was leading to connection failures in the test.

Thanks for your help on this!

**Comment by gsauthof** [comment]
Yes, that patch fixes it on Fedora, as well.

**Comment by ronf** [comment]
Terrific - thanks so much!

I've gone ahead and committed it to the the "develop" branch as commit f340747.

I also made another small change to help reduce the time it takes for a couple of the other tests, and this is available as commit bbb9fef. On my Linux system here, it shaves off 4-5 minutes of testing time (from 276 seconds to 15 seconds for the full test of tests in test_forward.py).

**Comment by ronf** [comment]
This change is now available in AsyncSSH 2.10.0.
